math_utils.py

```diff
@@ -1,23 +1,39 @@
 from scipy.spatial.transform import Rotation
 
-def quaternion_to_pitch_and_roll(x, y, z, w):
-    quaternion = Rotation.from_quat([x, y, z, w])
-    eulerAngles = quaternion.as_euler('xzy', degrees=True)
-
-    pitch, roll, _ = eulerAngles
+last_pitch = 0
+safety_counter = 0
+safety_mode = False
 
-    # pitch = max(-15.5, min(pitch, 33.75))
-    # roll = max(-40, min(roll, 40))
+def quaternion_to_pitch_and_roll(x, y, z, w):
+    global last_pitch, safety_counter, safety_mode
 
+    quaternion = Rotation.from_quat([x, y, z, w])
+    euler_angles = quaternion.as_euler('xzy', degrees=True)
+
+    pitch, roll, _ = euler_angles
+    
+    if (pitch > 0 and pitch-last_pitch >= 0.5): # Will have to be changed with testing. "0.5" should be replaced with the large single-frame delta seen in the test where an interlock occurred.
```

Comment:
Question: Why do we need to check if the pitch is positive? Can't the interlock also happen if the simulator moves downwards really quickly? I was thinking of logic like this instead:

```
delta_pitch = abs(pitch - last_pitch)
if (delta_pitch >= 0.5):
    ...
```
What do you think?

Reply:
It seems Rob was only really worried about interlocks for a large positive pitch. I think this may be the axis with the most extreme movement and also getting stuck at a large positive pitch can allow the door to slide open, which can damage parts and we are trying to avoid.